New climate component

quinnhosler opened the pull request:
adds support for climate entity, logic based off original zwave component. Requires [python-openzwave-mqtt #65](https://github.com/cgarwood/python-openzwave-mqtt/pull/65) to work as the value is a combined list/dict, which was previously not supported. 

Currently works well with my thermostat of a Honeywell T6 Pro. However, I did not use the previous zwave component with a climate device!

quinnhosler committed:
climate support added

quinnhosler committed:
formatting

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+    "furnace": HVAC_MODE_HEAT,
+    "fan only": HVAC_MODE_FAN_ONLY,
+    "dry air": HVAC_MODE_DRY,
+    "moist air": HVAC_MODE_DRY,
```
moist is the opposite of dry.

kpine commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+    "auto changeover": HVAC_MODE_HEAT_COOL,
+}
+
+MODE_SETPOINT_MAPPINGS = {
```
You should replace the strings with the integer values. This means it's unnecessary to account for any custom values in the XML files.

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+        self._current_swing_mode = None
+        self._swing_modes = None
+        self._unit = None
+        self._zxt_120 = None
```
Is this workaround still necessary or should that live in openzwave?

marcelveldt replied:
please leave all workarounds out because most of it has been addressed upstream. When we get issues reported we will look into that and see if it can be fixed upstream, if not a workaround in this code needs to be done.

In short: Do not blindly copy workarounds from the old implementation because the new implementation uses OZW 1.6 which addresses most of those shortcomings,

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+
+    def _update_operation_mode(self):
+        """Update hvac and preset modes."""
+        if self._mode():
```
Please use guard clauses and return as early as possible.

```python
if not self._mode():
    return
```

kpine commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+
+DEVICE_MAPPINGS = {REMOTEC_ZXT_120_THERMOSTAT: WORKAROUND_ZXT_120}
+
+HVAC_STATE_MAPPINGS = {
```
These should also use the integer values, not strings.

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+                None,
+            )
+
+            if _hvac_temp is None:
```
```python
if _hvac_temp is not None:
    # The current mode is a hvac mode
    self._hvac_mode = _hvac_temp
    self._preset_mode = PRESET_NONE
    return
```

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+
+    def _update_current_temp(self):
+        """Update current temperature."""
+        if self.values.temperature:
```
Same. Guard clause.

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+
+    def _update_fan_mode(self):
+        """Update fan mode."""
+        if self.values.fan_mode:
```
Same. Guard clause.

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+        self._target_temperature = None
+        self._target_temperature_range = (None, None)
+        if len(current_setpoints) == 1:
+            (setpoint,) = current_setpoints
```
To keep it more readable. 

```suggestion
            setpoint = current_setpoints[0]
```

balloob commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,545 @@
+
+    def _update_swing_mode(self):
+        """Update swing mode."""
+        if self._zxt_120 == 1:
```
Guard clause.

```python
if self._zxt_120 != 1 or not self.values.zxt_120_swing_mode:
    return
```

marcelveldt commented:
Hi @quinnhosler thanks for helping us out with your PR!
Please see some of the comments we gave, especially regarding hardcoded string values. Please prefer the ID's at all times because the labels are dynamic. 

Use a lookup table to retrieve the value for the ID or to convert the OZW ID (not the string label) to the Home Assistant representation.

quinnhosler committed:
refactoring

quinnhosler commented:
Alright, I took a look at all the corrections and suggestions and refactored the code. 

I removed anything that appeared to be a workaround, which included the ``aux_heat`` and ``swing_mode`` support. They appeared to only support a specific device, and as I do not have a device that supports these functions, I am unable to test them. They have been removed completely. 

Additionally, I did my best to use the OZW 1.6 mode integer values for the backend the labels are only used on the frontend. There is one spot where this was not possible, and that is regarding the ``HVAC_CURRENT_MAPPINGS`` constant. It appears OZW or qt-ozw does not send integer values and only labels. We can update this later as support is added. 

Finally, I have no idea if the ``ZWaveClimateMultipleSetpoint`` works. Again, my device does not support it and I have no way to test it. The code is mostly left untouched from the previous component. If anyone knows more about this, I'd be happy to know.

kpine commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -80,19 +54,9 @@
-    "heat_cool": ("setpoint_heating", "setpoint_cooling"),
 }
 
+# strings, OZW and/or qt-ozw does not send numeric values
```
This is not ideal... I can confirm, my tstat only shows string values. I've submitted a bug upstream, maybe it can be addressed.

https://github.com/OpenZWave/qt-openzwave/issues/65

marcelveldt replied:
Thanks @kpine for the issue report to OZW.
For now we'll go with the string values

marcelveldt commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+                "heat" in current_mode_label.lower()
+                and HVAC_MODE_HEAT in self._hvac_list
+            ):
+                self._hvac_mode = 0x01
```
possible to define a constant here ?

marcelveldt commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+                "cool" in current_mode_label.lower()
+                and HVAC_MODE_COOL in self._hvac_list
+            ):
+                self._hvac_mode = 0x02
```
and a const here too

marcelveldt commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        """Must know if single or double setpoint."""
```
please fix the docstring formatting

marcelveldt commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+    @property
+    def preset_modes(self):
+        """Return the list of available preset operation modes."""
+        """Need to be a subset of PRESET_MODES."""
```
please fix the docstring formatting

marcelveldt commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+    @property
+    def hvac_action(self):
+        """Return the current running hvac operation if supported."""
+        """Needs to be one of CURRENT_HVAC_*."""
```
please fix the docstring formatting

marcelveldt reviewed (changes requested):
Thanks for adjusting!
See the last feedback comments

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+            climate = ZWaveClimateMultipleSetpoint(values)
+        async_add_entities([climate])
+
+    async_dispatcher_connect(hass, "zwave_new_climate", async_add_climate)
```
Store the return value here in `hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE]`.
```suggestion
    hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE].append(
        async_dispatcher_connect(hass, "zwave_new_climate", async_add_climate)
    )
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from .const import DOMAIN
```
```suggestion
from .const import DATA_UNSUBSCRIBE, DOMAIN
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+
+    def __init__(self, values):
+        """Initialize the Z-Wave climate device."""
+        ZWaveDeviceEntity.__init__(self, values)
```
```suggestion
        super().__init__(values)
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+        self.update()
+        super().value_changed(value)
+
+    def update(self):
```
Remove this. It's not needed.

quinnhosler replied:
Which piece, the ``update`` function, the ``values_changed`` function, the call to ``super()`` or all of the above? 

I included the ``values_changed`` method here as there was no mechanism to update entity after a command was issued or a value changed. I think we may be able to switch from polling to either ``async_update`` or even ``async_write_ha_state`` or ``schedule_update_ha_state``. 

Lots of choices, and I'm not sure the which direction people were intending this to go.

MartinHjelmare replied:
Just remove the `update` method.

MartinHjelmare replied:
The integration is not polling. We subscribe to updates from the mqtt broker.

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+    @callback
+    def value_changed(self, value):
+        """Update after value change."""
+        self.update()
```
```suggestion
        self.update_properties()
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+        self._hvac_value_label_mapping = {}
+        self._hvac_label_value_mapping = {}
+        values_list = self._mode().value.get(VALUE_LIST)
+        mode_values = list(map(lambda x: x.get(VALUE_ID), values_list))
```
Please use a list comprehension instead of map + lambda.
```suggestion
        mode_values = [value.get(VALUE_ID) for value in values_list]
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+        self._fan_value_label_mapping = {}
+        self._fan_label_value_mapping = {}
+        self._fan_modes = []
+        for entry in self.values.fan_mode.value.get(VALUE_LIST):
```
Please use `dict[key]` when we know that the key is in the dict.
```suggestion
        for entry in self.values.fan_mode.value[VALUE_LIST]:
```

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+            return TEMP_CELSIUS
+        if self._unit == "F":
+            return TEMP_FAHRENHEIT
+        return self._unit
```
What other unit than Celsius or Fahrenheit would this be?

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+        """Return the highbound target temperature we try to reach."""
+        return self._target_temperature_range[1]
+
+    def set_temperature(self, **kwargs):
```
We should use the async api in the whole platform.
```suggestion
    async def async_set_temperature(self, **kwargs):
```

quinnhosler replied:
Is it planned to convert all the various components to async in the future? I don't think the other components are async. I'm also unfamiliar with the ``send_value`` method and it asynchronicity.

MartinHjelmare replied:
What components do you refer too? If you look at the other existing platforms in this integration they all use the async api for the actuator methods etc.

MartinHjelmare replied:
Hmm, it seems the fan platform doesn't do that. We should fix that (not here).

MartinHjelmare replied:
`send_value` is a callback that should run in the event loop since the `send_message` callback should run in the event loop.

MartinHjelmare replied:
Fixed the fan platform here:
https://github.com/cgarwood/homeassistant-zwave_mqtt/pull/107

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+            climate = ZWaveClimateSingleSetpoint(values)
+        elif values.primary.command_class == CommandClass.THERMOSTAT_MODE:
+            climate = ZWaveClimateMultipleSetpoint(values)
+        async_add_entities([climate])
```
Adding a list with `None` doesn't work.

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,448 @@
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Z-Wave Cover from Config Entry."""
```
```suggestion
    """Set up Z-Wave Climate from Config Entry."""
```

MartinHjelmare commented:
Please rebase on latest master branch and solve the merge conflict.

quinnhosler committed:
Apply suggestions from code review

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

quinnhosler committed:
additional revision

quinnhosler committed:
new constants

MartinHjelmare commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -57,11 +58,16 @@ async def mark_platform_loaded(platform):
 
-        hass.data[DOMAIN][entry.entry_id]["unsubscribe"] = await mqtt.async_subscribe(
-            hass, f"{TOPIC_OPENZWAVE}/#", async_receive_message
+        hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE].append(
```
This change has already been merged in master. Something is weird with this branch now.

MartinHjelmare replied:
Yeah, there are unrelated commits in this branch. Please *rebase* on latest master branch and remove any non relevant commits from your branch.

quinnhosler replied:
Will do, I'll work on it

quinnhosler replied:
I think that accomplished what we wanted, but let me know. Obviously I'm less familiar with certain parts of git

MartinHjelmare replied:
Good!

quinnhosler committed:
removed update method

quinnhosler committed:
async methods

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,461 @@
+        if not self.values.operating_state:
+            return
+        mode = self.values.operating_state.value
+        self._hvac_action = HVAC_CURRENT_MAPPINGS.get(str(mode).lower(), mode)
```
We should set `hvac_action` to `None` if the mapping is missing the mode. Ie remove the default in `dict.get`.

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,461 @@
+        self._current_temperature = None
+        self._hvac_action = None
+        self._hvac_list = None
+        self._hvac_mode = None
```
We should rename `self._hvac_mode` to `self._zw_hvac_mode` to make it clear that it's a zwave value and not a home assistant value.

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,461 @@
+        self._fan_label_value_mapping = {}
+        self._fan_modes = []
+        for entry in self.values.fan_mode.value[VALUE_LIST]:
+            self._fan_value_label_mapping[entry.get(VALUE_ID)] = entry.get(VALUE_LABEL)
```
Can the looked up entry id be `None`? Same question also for the looked up entry label.

Is there a risk that the entry doesn't have those id or label?

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,461 @@
+        self._preset_list = []
+        self._hvac_value_label_mapping = {}
+        self._hvac_label_value_mapping = {}
+        values_list = self._mode().value.get(VALUE_LIST)
```
Do we know that `VALUE_LIST` always will be in the mode value? If so we should access it with `self._mode().value[VALUE_LIST]`.

If not, we need to guard from setting `values_list` to `None`. We're expecting it to be a list below.

quinnhosler replied:
I'm erroring on the side of it always being available. I think it should be assuming proper implementation from OZW. If it were to change, then many parts of this would likely need to updated.

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,461 @@
+            return None
+        fan_mode_labels = []
+        for mode_value in self._fan_modes:
+            fan_mode_labels.append(self._fan_value_label_mapping.get(mode_value))
```
We know that the value label mapping has all the fan mode values, since we've set this ourselves earlier. So we should use `dict[key]` when looking up the mapping.

MartinHjelmare reviewed (commented):
Please go through all the dictionary look ups and check if we know that the key is present or not. Only use `dict.get` if we don't know that the key will be present in the dict. Otherwise we should use `dict[key]`.

quinnhosler committed:
change to self._zw_hvac_mode

quinnhosler committed:
.get(key) to [key]

quinnhosler committed:
constant, named arg, value_changed fixes

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,460 @@
+        self._unit = None
+        self.update_properties()
+
+    @callback
+    def _value_changed(self, value):
+        """Update after value change."""
+        self.update_properties()
+        super()._value_changed(value)
```
We recently changed our api here.
```suggestion
    @callback
    def on_value_update(self):
        """Update after value change."""
        self.update_properties()
```

quinnhosler replied:
okay, I was wondering why it had broken on my side

MartinHjelmare commented on custom_components/zwave_mqtt/climate.py:
```diff
@@ -0,0 +1,460 @@
+        """Set new target hvac mode."""
+        if not self._mode():
+            return
+        hvac_mode_value = self._hvac_label_value_mapping.get(hvac_mode.lower())
```
`hvac_mode` is a home assistant mode. I don't believe we have set the keys in `self._hvac_label_value_mapping` with home assistant modes, but with z-wave labels.

quinnhosler replied:
Yes, and I ran into a slight issue with this. The two solutions I can think of are have another dict that maps homeassistant labels to zwave labels, which we then map to zwave values. Second is to ignore any zwave labels that exist for predefined homeassistant modes.

Is there a solution you prefer?

MartinHjelmare replied:
Could we have another dict that is almost the reverse of `HVAC_MODE_MAPPINGS`?

quinnhosler replied:
Yeah, I think that makes sense. It would also let us get rid of ZW_HVAC_MODE_HEAT and ZW_HVAC_MODE_COOL in the process

MartinHjelmare commented:
Two comments, otherwise looks good.

quinnhosler committed:
HA to ZW constant, updated api

MartinHjelmare reviewed (approved):
Looks great!

marcelveldt commented:
Great, thanks!

marcelveldt commented:
Merged! A new version will be released to HACS in a bit, allowing some users to give this a testdrive!

joshtbernstein commented:
Please let me know if this comment should be somewhere else. Previously zwave climate devices would show up as multiple climate entities (for me both a heating and cooling). With a recent version of HA, this was changed to the more traditional single climate entity. I've just switched to this integration and noticed that I now have two climate entities for my zwave climate devices. Was this expected/intended behavior?

marcelveldt commented:
@joshtbernstein please open up a new issue for this report.
While replying: I'd like to investigate your report. Can you create an mqtt dump for me please and provide it with the issue report ?
In Hass --> developer tools --> services --> mqtt.dump

kpine commented:
I submitted #116 . Same problem here.
